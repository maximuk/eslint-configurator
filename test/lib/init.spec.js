const { BASE_CONFIG_NAME, PLUGINS_LIST } = require('../../lib/const.js');

const mockConfig = class {
  constructor(name) {
    this.name = name;
    this.configs = { [name]: name };
  }
};

jest.mock('inquirer');
jest.mock('../../lib/config.js', () => mockConfig);
jest.mock('../../lib/utils.js');

const inquirer = require('inquirer');
const utils = require('../../lib/utils.js');
const init = require('../../lib/init.js');

describe('./lib/init.js', () => {
  let answers;
  let selectedConfigs;

  beforeEach(() => {
    mockConfig.prototype.save = jest.fn();

    inquirer.prompt = jest.fn().mockReturnValueOnce({
      then: (callback) => {
        inquirer.prompt = inquirer.prompt
          .mockReturnValue({
            then: callback => callback(selectedConfigs),
          });
        callback(answers);
      },
    });

    answers = {
      ecmaFeatures: ['ecmaFeatures'],
      ecmaVersion: '7',
      env: ['env'],
      parser: 'parser',
      plugins: PLUGINS_LIST,
      sourceType: true,
    };
    selectedConfigs = PLUGINS_LIST.reduce((result, plugin) => {
      result[plugin] = plugin;

      return result;
    }, { [BASE_CONFIG_NAME]: BASE_CONFIG_NAME });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should create configs and format files', () => {
    init();

    expect(inquirer.prompt.mock.calls).toMatchSnapshot();
    expect(mockConfig.prototype.save.mock.calls).toMatchSnapshot();

    expect(utils.formatFiles).toHaveBeenCalledTimes(1);
  });

  it('should save with sourceType = script', () => {
    answers.sourceType = false;
    answers.plugins = [];
    selectedConfigs = { [BASE_CONFIG_NAME]: BASE_CONFIG_NAME };

    init();

    expect(inquirer.prompt.mock.calls).toMatchSnapshot();
    expect(mockConfig.prototype.save.mock.calls).toMatchSnapshot();

    expect(utils.formatFiles).toHaveBeenCalledTimes(1);
  });
});
