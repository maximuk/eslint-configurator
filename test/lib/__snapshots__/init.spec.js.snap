// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`./lib/init.js should create configs and format files 1`] = `
Array [
  Array [
    Array [
      Object {
        "choices": Array [
          "3",
          "5",
          "6",
          "7",
          "8",
          "9",
        ],
        "default": "7",
        "message": "Specify the version of ECMAScript syntax you want to use:",
        "name": "ecmaVersion",
        "type": "list",
      },
      Object {
        "default": true,
        "message": "Are you using ES6 modules?",
        "name": "sourceType",
        "type": "confirm",
      },
      Object {
        "choices": Array [
          "globalReturn",
          "impliedStrict",
          "jsx",
          "experimentalObjectRestSpread",
        ],
        "default": Array [
          "jsx",
          "experimentalObjectRestSpread",
        ],
        "message": "Which additional language features you’d like to use?",
        "name": "ecmaFeatures",
        "type": "checkbox",
      },
      Object {
        "choices": Array [
          "espree",
          "esprima",
          "babel-eslint",
        ],
        "default": "babel-eslint",
        "message": "Which parser are you going to use?",
        "name": "parser",
        "type": "list",
      },
      Object {
        "choices": Array [
          "builtin",
          "browser",
          "node",
          "commonjs",
          "shared-node-browser",
          "worker",
          "amd",
          "mocha",
          "jasmine",
          "jest",
          "phantomjs",
          "jquery",
          "qunit",
          "prototypejs",
          "shelljs",
          "meteor",
          "mongo",
          "protractor",
          "applescript",
          "nashorn",
          "serviceworker",
          "atomtest",
          "embertest",
          "webextensions",
          "es6",
          "greasemonkey",
        ],
        "default": Array [],
        "message": "Which env(s) are you going to use?",
        "name": "env",
        "pageSize": 10,
        "type": "checkbox",
      },
      Object {
        "choices": Array [
          "import",
          "react",
          "jsx-a11y",
        ],
        "default": Array [],
        "message": "Which plugins are you going to use?",
        "name": "plugins",
        "type": "checkbox",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          "No",
          "base",
        ],
        "default": "No",
        "message": "Do you want to use any base config?",
        "name": "base",
        "type": "list",
      },
      Object {
        "choices": Array [
          "No",
          "import",
        ],
        "default": "No",
        "message": "Do you want to use any import config?",
        "name": "import",
        "type": "list",
      },
      Object {
        "choices": Array [
          "No",
          "react",
        ],
        "default": "No",
        "message": "Do you want to use any react config?",
        "name": "react",
        "type": "list",
      },
      Object {
        "choices": Array [
          "No",
          "jsx-a11y",
        ],
        "default": "No",
        "message": "Do you want to use any jsx-a11y config?",
        "name": "jsx-a11y",
        "type": "list",
      },
    ],
  ],
]
`;

exports[`./lib/init.js should create configs and format files 2`] = `
Array [
  Array [
    "base",
    Object {
      "env": Object {
        "env": true,
      },
      "parser": "parser",
      "parserOptions": Object {
        "ecmaFeatures": Object {
          "ecmaFeatures": true,
        },
        "ecmaVersion": 7,
        "sourceType": "module",
      },
    },
  ],
  Array [
    "import",
    Object {},
  ],
  Array [
    "react",
    Object {},
  ],
  Array [
    "jsx-a11y",
    Object {},
  ],
]
`;

exports[`./lib/init.js should save with sourceType = script 1`] = `
Array [
  Array [
    Array [
      Object {
        "choices": Array [
          "3",
          "5",
          "6",
          "7",
          "8",
          "9",
        ],
        "default": "7",
        "message": "Specify the version of ECMAScript syntax you want to use:",
        "name": "ecmaVersion",
        "type": "list",
      },
      Object {
        "default": true,
        "message": "Are you using ES6 modules?",
        "name": "sourceType",
        "type": "confirm",
      },
      Object {
        "choices": Array [
          "globalReturn",
          "impliedStrict",
          "jsx",
          "experimentalObjectRestSpread",
        ],
        "default": Array [
          "jsx",
          "experimentalObjectRestSpread",
        ],
        "message": "Which additional language features you’d like to use?",
        "name": "ecmaFeatures",
        "type": "checkbox",
      },
      Object {
        "choices": Array [
          "espree",
          "esprima",
          "babel-eslint",
        ],
        "default": "babel-eslint",
        "message": "Which parser are you going to use?",
        "name": "parser",
        "type": "list",
      },
      Object {
        "choices": Array [
          "builtin",
          "browser",
          "node",
          "commonjs",
          "shared-node-browser",
          "worker",
          "amd",
          "mocha",
          "jasmine",
          "jest",
          "phantomjs",
          "jquery",
          "qunit",
          "prototypejs",
          "shelljs",
          "meteor",
          "mongo",
          "protractor",
          "applescript",
          "nashorn",
          "serviceworker",
          "atomtest",
          "embertest",
          "webextensions",
          "es6",
          "greasemonkey",
        ],
        "default": Array [],
        "message": "Which env(s) are you going to use?",
        "name": "env",
        "pageSize": 10,
        "type": "checkbox",
      },
      Object {
        "choices": Array [
          "import",
          "react",
          "jsx-a11y",
        ],
        "default": Array [],
        "message": "Which plugins are you going to use?",
        "name": "plugins",
        "type": "checkbox",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          "No",
          "base",
        ],
        "default": "No",
        "message": "Do you want to use any base config?",
        "name": "base",
        "type": "list",
      },
    ],
  ],
]
`;

exports[`./lib/init.js should save with sourceType = script 2`] = `
Array [
  Array [
    "base",
    Object {
      "env": Object {
        "env": true,
      },
      "parser": "parser",
      "parserOptions": Object {
        "ecmaFeatures": Object {
          "ecmaFeatures": true,
        },
        "ecmaVersion": 7,
        "sourceType": "script",
      },
    },
  ],
]
`;
